plugin/kubernetes: use transfer plugin

chrisohaver opened the pull request:
### 1. Why is this pull request needed and what does it do?

Allow use of the _transfer_ plugin in _kubernetes_.

The existing transfer option is refactored, but still functional - to be removed when the existing transfer option is eventually removed.

This PR also removes the zone transfer interface from the backend, since only k8s actually used it.  Etcd is mildly refactored to remove the no-op zone transfer stub.

### 2. Which issues (if any) are related?

### 3. Which documentation changes (if any) need to be made?

included.

### 4. Does this introduce a backward incompatible change or deprecation?

No, the existing transfer option is still functional in this PR.

chrisohaver commented:
Travis tests passed, despite the "Pending" status.  I guess the status update from travis to github failed.

miekg commented:
@chrisohaver does it make sense to do all transfer capable plugins in 1 go and then release 1.7.0? From fixing/tweaks rolling in, it seems we can do one (or two?) more 1.6.x releases and then tackle this?

chrisohaver commented:
If you'd like to defer merging this in favor of getting bug fixes merged, that makes sense.

... especially since the migration (this PR) doesn't add new functionality.

The k8s_external PR (#3419) _does_ add new functionality.  It didn't handle zone transfers before, and IMO the use case for doing zone transfers on an external zone is more practical than zone transfers on the internal cluster ip space.  But it's still new functionality, and therefore less urgent than bug fixes.

How do you want to handle the Notify addition to the transfer plugin interface (#3420)?  Review and merge now? Or wait until 1.7.0 or later.  I think its easiest to have that merged _before_ opening any PRs that depend on it (locally, I have notify implemented for k8s_external and kubernetes branches on ice until that merges).

miekg commented:
He Chris, sorry for lagging on this but busy with work. What would really help to get all this transfer stuff pushed forward is impl. of the auto plugin using transfer and it all works. If transfer is only useful for a subset of plugin we need to have good think if that's the best way forward.

chrisohaver commented:
> What would really help to get all this transfer stuff pushed forward is impl. of the auto plugin using transfer and it all works

OK - I did this [in a branch](https://github.com/chrisohaver/coredns/tree/auto-file-notify) in January when you requested it in the Notify PR... https://github.com/coredns/coredns/pull/3420#issuecomment-574735747

The commits that implement it are 3efabed3372d7792aee3150a229383c91bc02295 and c1d2623b775feeba01f4b4212316907a74bf79b2

The branch implements both transfer and notify, so it's dependent on the notify PR #3420.

Can we ahead and rebase/merge #3420 now, so I can push the branch up as a PR as is? 
Or do you want to see file/auto implemented _without_ notify first.

chrisohaver commented:
lol - looking back on that branch, I see i have a bunch of unintended changes of "notify" -> "notifyChan" in code comments and error messages.  I'll clean that up.

chrisohaver commented:
>I see i have a bunch of unintended changes of "notify" -> "notifyChan" in code comments and error messages. I'll clean that up.

OK, I cleaned that up and squashed to 0d6b91c0ae7511f000ce23f81c572cb801d1c1dc

chrisohaver commented:
@miekg, I went a head and created PR #3732 for just the transfer implementation without notify, since that #3420 has not yet merged.

chrisohaver commented:
ugh - more conflicts.

stp-ip commented on plugin/kubernetes/handler_test.go:
```diff
@@ -525,7 +524,7 @@ func (APIConnServeTest) Run()                                      {}
 func (APIConnServeTest) EpIndexReverse(string) []*object.Endpoints { return nil }
 func (APIConnServeTest) SvcIndexReverse(string) []*object.Service  { return nil }
-func (APIConnServeTest) Modified() int64                           { return time.Now().Unix() }
+func (APIConnServeTest) Modified() int64                           { return 2 }
```
What's the specific reason for this change?

chrisohaver replied:
I don't recall for certain. I suspect so that the serial is static during the tests, not changing based on when the test is run, but I don't see a change that adds a Serial check, so it could be extraneous.

chrisohaver replied:
Hmm ... I suspect it's part of the notify implementation, (which I had not posted because it depends on code that wasnt merged yet: #3420). 
Last November, I had Transfer+Notify working for both kubernetes and k8s_external. This may have snuck in from that.

stp-ip replied:
Generally would say nit, which shouldn't be in the PR. Doesn't break or add anything so also ok to keep it.

chrisohaver replied:
Actually - it may be directly related to this change.  There are tests here for incremental transfers, which check the serial.   Making the initial serial a low static number may have been to induce a "serial not changed" scenario to verify the IXFR short response when the zone is not changed.

But looking at the IXFR tests now, I don't see how they are actually working. On the surface, they look invalid. Still re-catching my bearings.

Sorry, it's been so long I don't have a crisp memory of the work.

chrisohaver replied:
OK, ... I see, the IXFR tests _are_ valid, and the "2" is there for the purpose of testing the IXFR fallback vs empty incremental response.  So it _is_ directly related to this change.

chrisohaver replied:
I'll add a comment to the code to explain it.

stp-ip replied:
Ahh now it makes sense. Thanks for bringing me up to speed.

Was missing it being used as the empty initialization and then check against hardcoded serials via k.Transfer().

All good.

miekg commented on plugin/kubernetes/handler.go:
```diff
@@ -29,7 +29,7 @@ func (k Kubernetes) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.M
 	switch state.QType() {
 	case dns.TypeAXFR, dns.TypeIXFR:
-		k.Transfer(ctx, state)
+		k.TransferDeprecated(ctx, state)
```
not worth the effort imo. I *think* the entire case can be dropped and the plugin does the right thing

miekg commented on plugin/kubernetes/transfer.go:
```diff
@@ -0,0 +1,216 @@
+
+// Transfer implements transfer.Transferer
+func (k *Kubernetes) Transfer(zone string, serial uint32) (<-chan []dns.RR, error) {
+	ch := make(chan []dns.RR, 2)
```
make the channel later, so you don't need close it if the zone isn't found

miekg replied:
I think in my PR I can get away with unbuffered and a go routine here. https://github.com/coredns/coredns/pull/3972/files#diff-f1591d056edea707afdeae51b8dc8b8fR25 onward

miekg commented on plugin/kubernetes/transfer.go:
```diff
@@ -0,0 +1,216 @@
+		close(ch)
+		return nil, errors.New("could not create soa")
+	}
+	soa, ok := soas[0].(*dns.SOA)
```
why does this need to be checked?

miekg commented on plugin/kubernetes/transfer.go:
```diff
@@ -0,0 +1,216 @@
+	// reverse zone transfers are not implemented
+	if rev := (plugin.Zones{"in-addr.arpa.", "ip6.arpa."}).Matches(zone); rev != "" {
+		close(ch)
+		return nil, transfer.ErrNotAuthoritative
```
technically we are authoritative to this is NotImp

miekg commented:
bunch of comments, but lgtm overall
